packages/shell-api/src/streams.ts

```diff
@@ -0,0 +1,172 @@
+          return v;
+        }
+        if (typeof prop === 'string') {
+          return target.getProcessor(prop);
+        }
+      },
+    });
+  }
+
+  constructor(private instanceState: ShellInstanceState) {
+    super();
+  }
+
+  get _mongo(): Mongo {
+    return this.instanceState.currentDb._mongo;
+  }
+
+  [asPrintable](): string {
+    return 'Atlas Stream Processing';
+  }
+
+  getProcessor(name: string) {
+    return new StreamProcessor(this, name);
+  }
+
+  @returnsPromise
+  async process(pipeline: Document[], options: Document = {}) {
+    if (!Array.isArray(pipeline) || !pipeline.length) {
+      throw new MongoshInvalidInputError(
+        'Invalid pipeline',
+        CommonErrors.InvalidArgument,
+        pipeline
+      );
+    }
+    const result = await this._runCommand({
+      processStreamProcessor: 1,
+      pipeline,
+      options,
+    });
+
+    if (result.ok !== 1) {
+      return result;
+    }
+
+    const { name, cursorId } = result as {
+      name: string;
+      limit: number;
+      cursorId: number;
+    };
+    const sp = this.getProcessor(name);
+
+    const stopAndDrop = () => {
+      sp.stop()
+        .then(() => sp.drop())
+        .catch(() => {
+          /* ignore */
+        });
+    };
+
+    try {
+      await sp._sampleFrom(cursorId);
+    } catch (err) {
+      // try to stop and drop the temp processor when interrupted
+      if (err instanceof MongoshInterruptedError) {
+        this._instanceState.messageBus.once(
+          'mongosh:interrupt-complete',
+          stopAndDrop
+        );
+      }
+
+      throw err;
+    }
+
+    // stop and drop the temp processor if reached the end of sample
+    return stopAndDrop();
+  }
+
+  @returnsPromise
+  async createStreamProcessor(
+    name: string,
+    pipeline: Document[],
+    options: Document
+  ) {
+    if (typeof name !== 'string' || name.trim() === '') {
+      throw new MongoshInvalidInputError(
+        `Invalid processor name: ${name}`,
+        CommonErrors.InvalidArgument
+      );
+    }
+    if (!Array.isArray(pipeline) || !pipeline.length) {
+      throw new MongoshInvalidInputError(
+        'Invalid pipeline',
+        CommonErrors.InvalidArgument,
+        pipeline
+      );
+    }
+    const result = await this._runCommand({
+      createStreamProcessor: name,
+      pipeline,
+      options,
+    });
+
+    if (result.ok !== 1) {
+      return result;
+    }
+
+    return this.getProcessor(name);
+  }
+
+  @returnsPromise
+  async listStreamProcessors(filter: Document) {
+    const result = await this._runCommand({
+      listStreamProcessors: 1,
+      filter,
+    });
+    if (result.ok !== 1) {
+      return result;
+    }
+    const processors = result.streamProcessors;
+    const sps = processors.map((sp: StreamProcessor) =>
+      this.getProcessor(sp.name)
+    );
+    return Object.defineProperty(sps, asPrintable, {
+      value() {
+        return JSON.stringify(processors, null, 2);
+      },
+    });
```

Comment:
Should this just be

```suggestion
    (sps as any)[asPrintable] = processors;
```

? Why the `JSON.stringify()` here?

Reply:
That was added so that the processor objects are printed in full. By default, deeply nested objects are not printed.

Reply:
`JSON.stringify()` is not the way to solve that problem, though … you could set a specific result type here (`sps[shellApiType] = '...'`) and then add it to the list here: https://github.com/mongodb-js/mongosh/blob/d5530affb0ccaace873c59f4ba6092df34f96fa8/packages/cli-repl/src/format-output.ts#L128